libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,564 @@
+
+  write_s()
+  dh_se()
+
+  # last payload must follow the ecrypted way of sending
+  var packed = packNoisePayload(p2psecret)
+  msg &= hs.ss.encryptAndHash(packed)
+
+  await conn.sendHSMessage(msg)
+  
+  let (cs1, cs2) = hs.ss.split()
+  return HandshakeResult(cs1: cs1, cs2: cs2, remoteP2psecret: remoteP2psecret, rs: hs.rs)
+
+proc handshakeXXInbound(p: Noise, conn: Connection, p2pProof: ProtoBuffer): Future[HandshakeResult] {.async.} =
+  const initiator = false
+
+  var
+    hs: HandshakeState
+    empty: seq[byte]
+  init hs.ss
+
+  var p2psecret = p2pProof.buffer
+
+  hs.ss.mixHash(p.commonPrologue)
+  hs.s.privateKey = p.noisePrivateKey
+  hs.s.publicKey = p.noisePublicKey
+
+  # -> e
+
+  var msg = await conn.receiveHSMessage()
+
+  read_e()
+
+  let earlyData = hs.ss.decryptAndHash(msg)
+
+  # <- e, ee, s, es
+
+  msg.setLen(0)
+
+  write_e()
+  dh_ee()
+  write_s()
+  dh_es()
+
+  var packedSecret = packNoisePayload(p2psecret)
+  msg &= hs.ss.encryptAndHash(packedSecret)
+
+  await conn.sendHSMessage(msg)
+
+  # -> s, se
+
+  msg = await conn.receiveHSMessage()
+
+  read_s()
+  dh_se()
+
+  var remoteP2psecret = hs.ss.decryptAndHash(msg)
+  unpackNoisePayload(remoteP2psecret)
+
+  let (cs1, cs2) = hs.ss.split()
+  return HandshakeResult(cs1: cs1, cs2: cs2, remoteP2psecret: remoteP2psecret, rs: hs.rs)
+
+proc readMessage(sconn: NoiseConnection): Future[seq[byte]] =
+  try:
+    return sconn.receiveEncryptedMessage()
+  except AsyncStreamIncompleteError:
+    trace "Connection dropped while reading"
+  except AsyncStreamReadError:
+    trace "Error reading from connection"
+
+proc writeMessage(sconn: NoiseConnection, message: seq[byte]): Future[void] =
+  try:
+    return sconn.sendEncryptedMessage(message)
+  except AsyncStreamWriteError:
+    trace "Could not write to connection"
+
+proc startLifetime*(sconn: NoiseConnection): Future[Connection] {.async.} =
+  proc writeHandler(data: seq[byte]) {.async, gcsafe.} =
+    trace "sending encrypted bytes", bytes = data.toHex()
+    await sconn.writeMessage(data)
+
+  var stream = newBufferStream(writeHandler)
+  asyncCheck readLoop(sconn, stream)
+  result = newConnection(stream)
+  result.closeEvent.wait().addCallback do (udata: pointer):
+    trace "wrapped connection closed, closing upstream"
+    if not isNil(sconn) and not sconn.closed:
+      asyncCheck sconn.close()
+
+proc handshake*(p: Noise, conn: Connection, initiator: bool): Future[NoiseConnection] {.async.} =
+  trace "Starting Noise handshake", initiator
+
+  # https://github.com/libp2p/specs/tree/master/noise#libp2p-data-in-handshake-messages
+  let
+    signedPayload = p.localPrivateKey.sign(PayloadString.getBytes & p.noisePublicKey.getBytes)
+    
+  var
+    libp2pProof = initProtoBuffer()
+    handshakeRes: HandshakeResult
+
+  libp2pProof.write(initProtoField(1, p.localPublicKey))
+  libp2pProof.write(initProtoField(2, signedPayload))
+  # data field also there but not used!
+  libp2pProof.finish()
+
+  if initiator:
```

Comment:
expression if again - shorter lifetime of handshakeRes makes code safer + let keeps it immutable:

```
let handshakeResult =
  if initiator: ...
  else: ...
```

Reply:
I had nearly forgotten this pattern worked in nim as well , honestly 😄

Reply:
it didn't, before :)